---
title: "curatedMetagenomicData curation report"
abstract: >
    This vignette checks every file under inst/curated that ends in `_metadata.tsv`
    against the template file [inst/extdata/template.csv](https://github.com/waldronlab/curatedMetagenomicDataCuration/blob/master/inst/extdata/template.csv), and shows problems found. Two types of errors can appear: 1) A "column name error" means a column name is not found at all in [template.csv](https://github.com/waldronlab/curatedMetagenomicDataCuration/blob/master/inst/extdata/template.csv). 2) An "entry error" means at least one value in a column does not match the rules defined for that column. This script will list the illegal values next to the row name they were found in. Entries are checked against the following rules: a) *uniqueness* - if "non-unique", then duplicate are allowed in the column, if "unique", then each duplicated value is an error and will be listed. b) *requiredness* - if "optional" then **NA** values are allowed, if "required" then any **NA** value is an error. c) *multiplevalues* - if TRUE, then multiple semicolon-separated are allowed in a single cell, if FALSE, no semicolon is allowed. d) *allowedvalues* is a regular expression definining legal values. These are assumed to be POSIX 1003.2 extended regular expressions (see `?grep`). Finally, the [template file](https://github.com/waldronlab/curatedMetagenomicDataCuration/blob/master/inst/extdata/template.csv) contains a free-form "description" field explaining what that column should contain, sometimes including units of measurement. The source repository for this document is https://github.com/waldronlab/curatedMetagenomicDataCuration.
output: 
  html_document:
    toc: true
vignette: >
    %\VignetteIndexEntry{curatedMetagenomicDataCuration}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache=FALSE)
library(curatedMetagenomicDataCuration)
```

This vignette was built on `r Sys.time()` based on `curatedMetagenomicDataCuration` version `r sessionInfo()$otherPkgs$curatedMetagenomicDataCuration$Version`.

Define base GitHub URLs for viewing / editing in place:
```{r}
baseurl <- "https://github.com/waldronlab/curatedMetagenomicDataCuration"
viewurl0 <- paste0(baseurl, "/tree/master/inst")
editurl0 <- paste0(baseurl, "/edit/master/inst")
```


If `firstonly <- TRUE`, only show results from the first dataset. This can be used as a temporary measure to shorten the output when troubleshooting the syntax-checking system itself. 
```{r}
firstonly <- FALSE
```

```{r template, echo=FALSE}
library(readr)
template <- read_csv(
  system.file("extdata/template.csv",
              package = "curatedMetagenomicDataCuration"),
  col_types = cols(
    col.name = col_character(),
    uniqueness = col_character(),
    requiredness = col_character(),
    multiplevalues = col_logical(),
    allowedvalues = col_character(),
    description = col_character()
  )
)
templatelist <- lapply(1:nrow(template), function(i){
  output <- t(template[i, ])
  output <- paste0("* *", rownames(output), "*: ", output[, 1])
  return( output )
})
names(templatelist) <- template$col.name
```

```{r dochecks, echo=FALSE, warning=FALSE}
allfiles <- dir(
  system.file("curated", package = "curatedMetagenomicDataCuration"),
  recursive = TRUE,
  pattern = "_metadata.tsv$",
  full.names = TRUE
)
allfiles.rel <- strsplit(allfiles, split="curatedMetagenomicDataCuration")
allfiles.rel <- sapply(allfiles.rel, function(x) x[[2]])
checks <- lapply(allfiles, function(fname){
  dat <- read.delim(fname, 
                    sep = "\t", 
                    stringsAsFactors = FALSE, 
                    na.strings = c("NA"),
                    check.names=FALSE)
  return( checkCuration(dat) )
})
names(checks) <- basename(allfiles)
```

```{r, results='asis', echo=FALSE}
ivec <- seq_along(allfiles)
if(firstonly) ivec <- 1

for (i in ivec) {
  datasetname <- names(checks)[i]
  viewurl <- paste0(viewurl0, allfiles.rel[i])
  editurl <- paste0(editurl0, allfiles.rel[i])
  cat("# ", datasetname, "\n")
  cat("\n")
  cat("[View](", viewurl, ") or [edit](", editurl, ") this file directly on github.com. \n")
  cat("\n")
  if(identical(checks[[i]], list(missingcols = NULL, invalidcols = NULL, values = NULL))){
    cat("All checks OK. \n")
    cat("\n")
    next
  }
  if(!is.null(checks[[i]]$missingcols)){
    cat("## Required columns that are missing \n")
    cat(checks[[i]]$missingcols, "\n")
    cat("\n")
  }
  if(!is.null(checks[[i]]$invalidcols)){
    cat("## Column name errors \n")
    for (j in seq_along(checks[[i]]$invalidcols)){
      cat(paste0("* \"", checks[[i]]$invalidcols[j], "\"", " is not defined in the template. \n"))
      cat("\n")
    }
  }
  cat("\n")
  if(!is.null(checks[[i]]$values)){
    cat("## Entry errors \n")
    cat("\n")
    for (j in seq_along(checks[[i]]$values)){
      if(!any(grepl("!!!", checks[[i]]$values[, j]))) next
      cat("\n")
      problemvariable <- colnames(checks[[i]]$values)[j]
      cat("### ", problemvariable, "\n")
      cat("\n")
      cat("**Template definition** \n")
      cat("\n")
      for (k in 2:6)
        cat(templatelist[[problemvariable]][k], "  \n ")
      cat("\n")
      cat("**Errors** \n")
      cat("\n")
      output <- paste0(checks[[i]]$values$sampleID, "   :   ", checks[[i]]$values[, j])
      for (k in seq_along(output))
        if(grepl("!!!", output[k]))
          cat(k, ". ", gsub("!!!", "\"", output[k]), " \n ")
    }
  }
  cat("\n")
}
```
